readme.md
# RT-Thread操作系统的μC/OS-III兼容层
## μC/OS-III RTOS Application Compatibility Layer (ACL) for RT-Thread
## 让基于μC/OS-III开发的应用层无感地迁移到RT-Thread操作系统

### 如果喜欢请Star，这是对本开源项目最大的鼓励，谢谢；如果想要贡献PR，请fork

--------------------

[TOC]

# 0 前排提示

本文含有图片，受限于**中国大陆互联网环境**，访问github时，**readme.md(本文件)的图片一般加载不出来**，因此我导出了.pdf文件。如果您需要详细阅读，可以将项目下载或clone下来，阅读[docs/中文说明文档.pdf](docs/中文说明文档.pdf)文件。






# 1 概述

这是一个针对RT-Thread国产操作系统的μCOS-III操作系统兼容层，可以让基于美国Micriμm公司的μCOS-III操作系统的项目快速、无感地迁移到RT-Thread操作系统上。在兼容层的设计、编写上尊重原版μC/OS-III，保证原版μC/OS-III的原汁原味。本项目100%由C语言完成。

支持版本：μC/OS-III 3.00-3.08全部版本

视频介绍&教程：https://www.bilibili.com/video/BV1b54y1U7AG



## 1.1 RT-Thread的其他RTOS兼容层

- RT-Thread操作系统的μCOS-II兼容层：https://github.com/mysterywolf/RT-Thread-wrapper-of-uCOS-II
- RT-Thread操作系统的RTX(即CMSIS-RTOS)兼容层：https://github.com/RT-Thread-packages/CMSIS_RTOS1
- RT-Thread操作系统的RTX5(即CMSIS-RTOS2)兼容层：https://github.com/RT-Thread-packages/CMSIS_RTOS2



## 1.2 本兼容层适合于

- 之前学习过μCOS-III操作系统，意图转向学习RT-Thread国产操作系统。本兼容层可以帮您用已有的μCOS-III编程经验和习惯快速将项目跑起来，日后在应用过程中深入熟悉RT-Thread的API函数，逐步向RT-Thread过度，降低您的学习门槛和时间成本。**有了本兼容层，对RT-Thread API以及编程风格的不熟悉再也不是您学习RT-Thread的阻力！**

- 现有任务（线程）模块采用μCOS-III编写，想要用在基于RT-Thread的工程上

- 老项目需要从μCOS-III操作系统向RT-Thread操作系统迁移

- 当需要快速基于RT-Thread开发产品，但是工程师之前均采用μC/OS开发，从未用过RT-Thread的开发经验。本兼容层可以帮助让工程师快速基于μC/OS-III开发经验开发产品，简化软件的重用、缩短微控制器新开发人员的学习过程，并缩短新设备的上市时间。

- 避免在从μCOS-III迁移到RT-Thread时，由于μCOS-III的编程经验导致的思维定式引发的错误，这种错误一般很难被发现 
  
     > ​    例如：
     >
     > 1. 两个操作系统对于任务/线程挂起、解挂函数的区别。   
     > RT-Thread不支持任务嵌套挂起、解挂   
     > ​μCOS-III支持任务嵌套挂起、解挂
     >
     > 2. 软件定时器参数的不同
     >
     > 3. 任务堆栈的数据类型不同
     
- 本兼容层实现了与Micriμm公司专门为其旗下产品μC/OS等开发的专用软件μC/Probe的对接，可以通过该软件以图像化形式查看、调试RT-Thread内核以及μCOS-III兼容层的相关信息



## 1.3 版本详细信息
|    组件名称    | 版本号  |                     配置文件                     |                   说明                    |
| :------------: | :-----: | :----------------------------------------------: | :---------------------------------------: |
| RT-Thread nano |  3.1.3  |                    rtconfig.h                    |                                           |
|   μC/OS-III    | 3.03.00 | os_cfg_app.h  <br />os_cfg.h<br />os_app_hooks.c | 兼容层兼容3.00.00-3.08.00全部μCOS-III版本 |
|     μC/CPU     | 1.30.00 |                    cpu_cfg.h                     |  部分升级到最新版本, 兼容所有Micriμm软件  |
|     μC/LIB     | 1.39.00 |                    lib_cfg.h                     |                 最新版本                  |



## 1.4 官网

RT-Thread：https://www.rt-thread.org/   
文档中心：https://www.rt-thread.org/document/site/tutorial/nano/an0038-nano-introduction/

μCOS-III：https://www.micrium.com/  
文档中心：https://doc.micrium.com/display/kernel304/uC-OS-III+Documentation+Home





# 2 使用

## 2.1 Keil-MDK仿真工程
本仿真工程是基于STM32F103平台。

Keil工程路径：[rt-thread-3.1.3/bsp/stm32f103/Project.uvprojx](rt-thread-3.1.3/bsp/stm32f103/Project.uvprojx)

需要提前安装好RT-Thread Nano-3.1.3 [Keil支持包](https://www.rt-thread.org/download/mdk/RealThread.RT-Thread.3.1.3.pack).

**注意：调试串口使用的是USART2，不是USART1**

<img src="docs/pic/usart2.png" alt="usart2"  />





## 2.2 迁移步骤

**（如果使用的是RT-Thread Nano版请参见以下步骤；若使用RT-Thread完整版可以直接跳转至[Env工具自动化配置到工程中](#6 Env工具自动化配置到工程中)章节）**

1. 将uCOS-III、uC-LIB、uC-CPU三个文件夹内的所有文件都加入到你的工程中，最好保持原有文件夹的结构。相较于原版μCOS-III增加了`os_rtwrap.c`文件，负责对RT-Thread和μCOS-III的转换提供支持。
3. 浏览一下`μCOS-III/os.h`文件，看一下错误代码，这个错误代码和原版μCOS-III是有一定区别的。  
   **注意: 请勿随意打开注释掉的错误码枚举体成员，** 如果用户使用到了这些注释掉的成员,则会在迁移时编译报错,用以提醒用户这些错误代码在兼容层已经不可用。
4. 配置`os_cfg.h`和`os_cfg_app.h`  
   每个选项的配置说明和原版μCOS-III一致，若有不同，我已经在注释中有所解释。  
   **原版μCOS-III配置**说明可参见：  
   a)《嵌入式实时操作系统μC/OS-III》北京航空航天大学出版社 宫辉等译 邵贝贝审校   
   b) Micriμm公司μCOS-III[在线文档](https://doc.micrium.com/display/kernel304/uC-OS-III+Features+os_cfg.h)
5. μCOS-III原版定时器回调函数是在定时器线程中调用的，而非在中断中调用，因此要使用μCOS-III兼容层的软件定时器，需要将rtconfig.h中的宏定义`RT_USING_TIMER_SOFT`置1。



## 2.3 os_cfg.h配置文件

 ```c
#define  OS_CFG_TMR_TASK_RATE_HZ 100u /* Rate for timers (100 Hz Typ.) */
 ```
在原版μCOS-III中，该宏定义定义了软件定时器的时基信号，这与RT-Thread的软件定时器有本质的不同，在RT-Thread中，软件定时器的时基信号就等于OS Ticks。因此为了能够将μCOS-III软件定时器时间参数转为RT-Thread软件定时器的时间参数，需要用到该宏定义。请使该宏定义与原工程使用μCOS-III时的该宏定义参数一致。需要注意的是，虽然在兼容层中定义了软件定时器的时基频率，但是在兼容层内部使用的RT-Thread软件定时器的时基频率等同于OS Ticks，因此`OS_TMR`结构体的`.Match`成员变量其保存的数值是以OS Ticks频率来计算的。



## 2.4 os_cfg_app.h配置文件

该文件仅保留了OS Tick频率的配置、定时器任务以及统计任务的配置。其他配置项本兼容层用不到（例如中断任务），予以删除。



## 2.5 运行

### 2.5.1 官方标准手动初始化流程

本兼容层完全兼容官方给出的标准初始化流程，如果您兼容老项目，μCOS-III初始化部分无需做任何修改。具体初始化流程代码参见工程`main.c`文件，参考文献参见 [docs/uCOS-III官方初始化流程.pdf](docs/uCOS-III官方初始化流程.pdf) 



### 2.5.2 最简手动初始化流程

最简初始化流程是指本兼容层的初始化流程，不必像官方给出的初始化流程一样复杂。如果您不是想要兼容已有老工程，而是新建一个工程的话，可以采用最简手动初始化流程：

```c
#include <os.h> /*头文件保持和原版μCOS-III相同*/

int main(void) /*RT-Thread main线程*/
{
    OS_ERR err;
    
    OSInit(&err);                                   /*uCOS-III操作系统初始化*/
    
    OSStart(&err);                                  /*开始运行uCOS-III操作系统*/
    
#if OS_CFG_APP_HOOKS_EN > 0u
    App_OS_SetAllHooks();                           /*设置钩子函数*/
#endif  
    
#if OS_CFG_STAT_TASK_EN > 0u
    OSStatTaskCPUUsageInit(&err);                   /*统计任务*/    
    OSStatReset(&err);                              /*复位统计数据*/    
#endif	  

}
```



### 2.5.3 自动初始化流程

如果您在应用层中不想手动初始化本兼容层，可以在`rtconfig.h`文件中定义`PKG_USING_UCOSIII_WRAPPER_AUTOINIT`宏定义。请参见 [6.2.1章节](###6.2.1 Enable uCOS-III wrapper automatically init)（**如无特殊要求，建议采用该种方式**）。



### 2.5.4 精简版兼容层

如果你在使用过程中不需要兼容任务/内核对象结构体的成员变量，或者不需要使用uC/Probe软件监控兼容层状态，可以在`rtconfig.h`文件中定义`PKG_USING_UCOSIII_WRAPPER_TINY`宏定义。请参见 [6.2.2章节](###6.2.2 Enable uCOS-III wrapper tiny mode)。



## 2.6 注意
1. μCOS-III的任务堆栈大小单位是`sizeof(CPU_STK)`，而RT-Thread的线程堆栈大小单位是`sizeof(rt_uint8_t)`，虽然在兼容层已经做了转换，但是在填写时一定要注意，所有涉及到μCOS-III的API、宏定义全部是按照μCOS-III的标准，即堆栈大小为`sizeof(CPU_STK)`，**切勿混搭**！这种错误极其隐晦，一定要注意！**下面是混搭的错误示例**：</br>

    ```c
    ALIGN(RT_ALIGN_SIZE)
    static rt_uint8_t thread2_stack[1024];//错误：混搭RT-Thread的数据类型定义线程堆栈
    
    OSTaskCreate(&thread2,
                 (CPU_CHAR*)"thread2",
                 thread2_entry,	
                 RT_NULL,
                 THREAD_PRIORITY,
                 thread2_stack,
                 sizeof(thread2_stack)/10,//任务堆栈深度限位(错误：这个参数的单位是sizeof(CPU_STK))
                 sizeof(thread2_stack),//任务堆栈大小(错误：这个参数的单位是sizeof(CPU_STK))
                 0,
                 THREAD_TIMESLICE,
                 0,
                 OS_OPT_TASK_STK_CHK|OS_OPT_TASK_STK_CLR,
                 &err);
    ```
    **下面是正确写法**：</br>

    ```c
    #define THREAD_STACK_SIZE       256 //正确，要通过宏定义单独定义堆栈大小，单位为sizeof(CPU_STK)
    ALIGN(RT_ALIGN_SIZE)
        static CPU_STK thread2_stack[THREAD_STACK_SIZE];//正确，使用uCOS-III自己的数据类型定义任务堆栈
    
    OSTaskCreate(&thread2,
                 (CPU_CHAR*)"thread2",
                 thread2_entry,
                 RT_NULL,
                 THREAD_PRIORITY,
                 thread2_stack,
                 THREAD_STACK_SIZE/10,//任务堆栈深度限位(正确)
                 THREAD_STACK_SIZE,//任务堆栈大小(正确)
                 0,
                 THREAD_TIMESLICE,
                 0,
                 OS_OPT_TASK_STK_CHK|OS_OPT_TASK_STK_CLR,
                 &err);
    ```

2. **切勿将同一个内核对象/线程的RT-Thread和μCOS-III的API混搭使用。**   
    例如RT-Thread中的`rt_thread_suspend` / `rt_thread_resume` 函数仅支持一次挂起/解挂；而μCOS-III的`OSTaskSuspend` / `OSTaskResume` 函数是支持嵌套挂起/解挂的，为此需要继承`struct rt_thread`结构体并在其基础上增加成员`.SuspendCtr`变量实现该功能。若采用`rt_thread_init`函数初始化线程，该函数并不会管理μCOS-III兼容层的成员变量，`.SuspendCtr`也不会创建和初始化，若此时调用`OSTaskSuspend` / `OSTaskResume`函数试图指向`.SuspendCtr`成员变量，将会访问非法内存地址(因为`rt_thread_init`初始化的线程`.SuspendCtr`成员变量根本不存在)！
   
   但是允许非同一个内核对象/线程的RT-Thread和μCOS-III的API混搭使用。例如可以使用`OSTaskCreate`函数创建任务，同时在该任务内部可以使用RT-Thread的信号量API进行操作。总之，如果你用μCOS-III的API创建内核对象或者线程，在后续对**该**内核对象或任务的操作必须沿用μCOS-III的API；反之，也必须沿用RT-Thread的API。
   
3. 兼容层取消了原版μCOS-III中的时间戳功能  
    在μCOS-III中，时间戳主要用于测量中断关闭时间，以及任务单次执行时间以及最大时间等涉及到精度较高的时长测量。该特性在μCOS-II以及RT-Thread中均没有，因此本兼容层不予实现。







# 3 接口
## 3.1 没有实现兼容的API（仅2个）

RT-Thread目前没有动态更改时间片大小的功能：

 ```c
void  OSTaskTimeQuantaSet (OS_TCB *p_tcb, OS_TICK time_quanta, OS_ERR *p_err);
 ```

兼容层取消原版μCOS-III中的多内核对象等待(Multi-Pend)功能，该功能在原版3.05.00版本开始向用户发出警告不要使用该功能(原文措辞为deprecated)，从3.06.00版本开始删除了该功能，因此本兼容层没有实现该函数的兼容：

```c
OS_OBJ_QTY  OSPendMulti (OS_PEND_DATA  *p_pend_data_tbl,
                         OS_OBJ_QTY     tbl_size,
                         OS_TICK        timeout,
                         OS_OPT         opt,
                         OS_ERR        *p_err);
```



## 3.2 钩子函数

**μCOS-III的钩子函数仅对μCOS-III兼容层负责。** 即如果你注册了`OSTaskDelHook`函数，他仅会在调用OSTaskDel函数时被调用，不会在调用`rt_thread_detach`函数时被调用(这个由RTT的钩子函数负责)。这样做是为了层次分明，防止μCOS-III兼容层插手RT-Thread内部事务。

μCOS-III的钩子函数在两个文件中实现：`os_cpu_c.c`和`os_app_hooks.c` 。按照μCOS-III的思想，`os_cpu_c.c`提供原始的钩子函数（即这些钩子函数被相应的函数直接调用），该文件以及其内部的钩子函数是移植工程师编写的内容，应用工程师不应该操作这个文件的内容，`os_cpu_c.c`文件的钩子函数提供相应的函数指针供`os_app_hooks.c`文件内的钩子函数注册和使用，这个文件内的钩子函数应用工程师是可以操作的。换句话说，我们有什么需要在钩子函数中调用的函数，应该放在`os_app_hooks.c`文件中。

以下原版μCOS-III钩子函数将予以取消，由RT-Thread接管相关钩子函数接管：

```c
void          OSTaskReturnHook          (OS_TCB *p_tcb);
void          OSTaskSwHook              (void);
void          OSTimeTickHook            (void);
```

​	同时，上述钩子函数对应的应用级钩子函数也被取消：

```c
void  App_OS_TaskReturnHook (OS_TCB  *p_tcb);
void  App_OS_TaskSwHook (void);
void  App_OS_TimeTickHook (void);
```



## 3.3 统计任务（OS_StatTask()、os_stat.c）

在μCOS-III中，统计任务是一个系统任务，通过`OS_CFG_STAT_TASK_EN`宏决定是否开启，可以在系统运行时做一些统计工作。例如统计总的CPU使用率（0.00% - 100.00%）、各任务的CPU使用率（0.00% - 100.00%）以及各任务的堆栈使用量。CPU的利用率用一个0-10000之间的整数表示（对应0.00% - 100.00%）。


changelog.md
# 维护日志

### 2020-06-29 

- 初步搭建框架
- 完成信号量兼容
- 完成`os_time.c`系统延时有关兼容


### 2020-06-30 

- 完成软件定时器兼容


### 2020-07-01  

- 实现互斥量兼容


### 2020-07-02  

- 实现消息队列兼容


### 2020-07-03  

- 实现信号标志组兼容

- 实现任务内建寄存器兼容

- 实现任务内建消息队列兼容

### 2020-07-04  

- 实现任务内建信号量兼容
- 实现uCOS-III安全认证检查相关兼容
- 对接`os_cfg.h`宏定义
- 增加`os_cfg_app.h`并与之对接
- 完成`os_sem.c` `os_q.c`文件宏定义检查以及流程的确认

### 2020-07-05  

- 完成`os_flag.c` `os_mutex.c` `os_tmr.c` `os_task.c` `os_time.c` `os_core.c`文件宏定义检查以及流程的确认
- 优化`os_cfg_app.h`、`os_cfg.h`

### 2020-07-06  

- 充实`readme.md`
- 实现`OSSemSet`、`OSTaskSemSet`函数
- 优化`OSTaskSuspend`、`OSTaskResume`函数，使其具备嵌套挂起/解挂能力
- 实现钩子函数

### 2020-07-07  

- 移植`os_mem.c`文件
- 优化了内核对象的结构体

### 2020-07-08  

- 实现`OSTmrStateGet`函数
- 构建统计任务框架
- 修改`OSTaskDel`函数没有删除任务内建信号量、任务内建队列的bug
- 移植`OS_TaskDbgListAdd`、`OS_TaskDbgListRemove`函数
- 实现`OS_TaskInitTCB`函数
- 实现Idle任务（实则为RTT Idle任务的回调函数）
- 实现Idle任务钩子函数`OSIdleTaskHook()`

### 2020-07-09

- 实现uCOS-III内部任务——统计任务，可以正确计算CPU使用率、每个任务的使用堆栈

### 2020-07-10

- 实现`OSTimeDlyResume`函数
- 修复`OSSemDel`函数无法返回当前等待信号量数量、无法使用`OS_OPT_DEL_NO_PEND`选项的问题
- 修复`OSMutexDel`函数无法返回当前等待信号量数量、无法使用`OS_OPT_DEL_NO_PEND`选项的问题
- 修复`OSQDel`函数无法返回当前等待信号量数量、无法使用`OS_OPT_DEL_NO_PEND`选项的问题
- 修复`OSFlagDel`函数无法返回当前等待信号量数量、无法使用`OS_OPT_DEL_NO_PEND`选项的问题

### 2020-07-11


- 增加兼容原版`OS_TCB`结构体成员变量
- 实现`OSSemPendAbort`函数，`OSSemPend`函数增加了处理`OS_ERR_PEND_ABORT`的功能

### 2020-07-12

- 新增`os_rtwrap.c`文件，负责对RT-Thread和μCOS-III的转换提供支持。并在`os.h`文件末尾增加`os_rtwrap.c`内函数的声明
- 实现`OSTaskSemPendAbort`函数
- 实现`OSTaskQPendAbort`函数
- 实现`OSSemPendAbort`函数，`OSSemPend`函数增加了处理`OS_ERR_PEND_ABORT`的功能
- 实现`OSSemMutexAbort`函数，`OSMutexPend`函数增加了处理`OS_ERR_PEND_ABORT`的功能
- 实现`OSSFlagPendAbort`函数，`OSFlagPend`函数增加了处理`OS_ERR_PEND_ABORT`的功能
- 实现`OSQPendAbort`函数，`OSQPend`函数增加了处理`OS_ERR_PEND_ABORT`的功能

### 2020-07-14

- 实现`OSSchedRoundRobinCfg`函数
- 实现`OSFlagPendGetFlagsRdy`函数
- 实现`OSQFlush`函数
- 实现`OSTaskQFlush`函数
- 补充`timer_sample.c`例程

### 2020-07-15

- 移植`cpu_core.c`文件增加`CPU_Init`函数，老项目如果用到该函数在迁移时不会报错，函数内部没有实际动作
- 优化了`timer_sample.c`例程
- 补充`mutex_sample.c`例程

### 2020-07-16

- 任务内建消息队列、任务内建信号量直接沿用任务的name，不再strcat字符串
- 实现`OSTmrDel`函数回调函数功能
- 增加`OSTmrDbgListPtr`链表
- 实现了由RT-Thread定时器回调函数一个参数，到μCOS-III定时器回调函数两个参数的转换，在移植时可以直接沿用原版回调函数，无需再修改为RTT定时器的回调函数样式
- 完整实现了`p_tmr->State`的所有选项，重新实现了`OSTmrStateGet`函数
- 实现μCOS-III软件定时器带有延迟的周期延时功能(`os_tmr.c`受限项全部解决)

### 2020-07-18

- 开始对接FinSH，实现msh命令

### 2020-07-19

- `OSInit`函数内加入临界区保护
- 实现`OSSemDbgListPtr`链表
- 实现`OSSemQty`全局变量
- 实现`OSMutexDbgListPtr`链表
- 实现`OSMutexQty`全局变量
- 实现`OSQDbgListPtr`链表
- 实现`OSQQty`全局变量
- 实现`OSFlagDbgListPtr`链表
- 实现`OSFlagQty`全局变量

### 2020-07-20

- 在`os_rtwrap.c`文件中增加`rt_ipc_post_all`和`rt_sem_release_all`函数，并实现了`OSSemPost`函数`OS_OPT_POST_ALL`选项

### 2020-07-22

- 增加`sem_sample.c`例程

### 2020-07-23

- 增加`q_sample.c`例程
- 修改`main.c`，严格按照μCOS-III官方给出的初始化流程进行初始化

### 2020-07-24

- 增加μC/LIB全部文件（V1.39.00）
- 更改工程结构，将工程文件提到一级目录
- 实现μC/CPU中`CPU_NameClr`、`CPU_NameGet`、`CPU_NameSet`函数

### 2020-07-25

- 增加`os_dbg.c`、`os_cfg_app.c`文件，开始与μC/Probe软件进行对接
- 实现`OS_TCB`成员变量`.TaskState`
- `OS_TCB`成员变量`.StkPtr`改为在统计任务中定时更新

### 2020-07-26

- 实现`OS_TCB`成员变量`.PendOn`、`.DbgName`
- 信号量、互斥量、消息队列、事件标志组结构体增加成员变量`.DbgName`

### 2020-07-27

- 修复`OSMutexPendAbort`、`OSSemPendAbort`、`OSQPendAbort`、`OSFlagPendAbort`函数返回值由当前还有多少任务在等待该内核对象 改为 放弃了多少个任务
- 增加`task_sem_sample.c`例程
- 实现`OS_MUTEX`结构体`.OwnerNestingCtr`、`.OwnerTCBPtr`、`.OwnerOriginalPrio`成员变量
- 修复了CPU使用率突然飙升至100%的问题，该问题为μCOS-III原版（3.03版本）内部bug，即便是最新版的3.08版本该问题依然没有得到彻底解决（仅仅是将飙升至100%改成了骤降至0%）

### 2020-07-28

- 增加`OS_CFG_TASK_PROFILE_EN`宏定义，用以降低资源占用
- 在`os_cfg_app.c`文件中增加`OS_CFG_DBG_EN`宏定义，用以降低资源占用

### 2020-07-29

- 完善readme

### 2020-08-03

- 增加`OS_TASK_STATE_SUSPENDED`任务状态

### 2020-08-07

- 将例程由sample字眼改成example
- 增加`PKG_USING_UCOSIII_WRAPPER_AUTOINIT`宏

### 2020-08-08

- 增加SConscript脚本，以支持env工具进行自动化添加软件包

### 2020-08-12

- 完善readme文档

### 2020-08-13

-  修复`OSTmrRemainGet`函数计算剩余时间的错误
-  实现`OS_TMR`结构体`.Match`和`.Remain`成员变量

### 2020-08-16

- 实现错误代码`OS_ERR_TASK_SUSPENDED`  、`OS_ERR_TASK_OPT`、`OS_ERR_LOCK_NESTING_OVF`、`OS_ERR_MUTEX_NESTING`、`OS_ERR_MUTEX_OWNER`、`OS_ERR_ROUND_ROBIN_DISABLED`、`OS_ERR_STK_LIMIT_INVALID` 、`OS_ERR_FLAG_PEND_OPT`
- 对`OSTaskStkChk`函数当传入`OS_TCB*`参数为空时，转换为当前`OS_TCB*`的逻辑做了调整

### 2020-08-17

- 对函数`OSTmrRemainGet`、`OSTmrStop`、`OSTmrStateGet`、`OSTmrStart`、`OSTmrDel`仿照原版增加定时器状态检查功能

### 2020-08-18

- 修复`OSInit`函数return分支没有调用`CPU_CRITICAL_EXIT()`打开中断的bug
- 增加视频介绍&教程链接
- 增加3.08版本中新增的`OSInitialized`全局变量；`OS_FALSE`、`OS_TRUE`宏定义；错误代码`OS_ERR_OS_NOT_INIT`、`OS_ERR_OS_NO_APP_TASK`

uC-CPU/cpu.h
/*
 * Copyright (c) 2021, Meco Jianting Man <jiantingman@foxmail.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Change Logs:
 * Date           Author       Notes
 * 2020-06-29     Meco Man     the first verion
 */
/*
*********************************************************************************************************
*                                               uC/CPU
*                                    CPU CONFIGURATION & PORT LAYER
*
*                    Copyright 2004-2020 Silicon Laboratories Inc. www.silabs.com
*
*                                 SPDX-License-Identifier: APACHE-2.0
*
*               This software is subject to an open source license and is distributed by
*                Silicon Laboratories Inc. pursuant to the terms of the Apache License,
*                    Version 2.0 available at www.apache.org/licenses/LICENSE-2.0.
*
*********************************************************************************************************
*/
/*
*********************************************************************************************************
*                                                uC/CPU
*                                    CPU CONFIGURATION & PORT LAYER
*
*                          (c) Copyright 2004-2013; Micrium, Inc.; Weston, FL
*
*               All rights reserved.  Protected by international copyright laws.
*
*               uC/CPU is provided in source form to registered licensees ONLY.  It is
*               illegal to distribute this source code to any third party unless you receive
*               written permission by an authorized Micrium representative.  Knowledge of
*               the source code may NOT be used to develop a similar product.
*
*               Please help us continue to provide the Embedded community with the finest
*               software available.  Your honesty is greatly appreciated.
*
*               You can find our product's user manual, API reference, release notes and
*               more information at https://doc.micrium.com.
*               You can contact us at www.micrium.com.
*********************************************************************************************************
*/

/*
*********************************************************************************************************
*
*                                            CPU PORT FILE
*
*                                            ARM-Cortex-M4
*                                            IAR C Compiler
*
* Filename      : cpu.h
* Version       : V1.30.00
* Programmer(s) : JJL
*                 BAN
*********************************************************************************************************
*/


/*
*********************************************************************************************************
*                                               MODULE
*
* Note(s) : (1) This CPU header file is protected from multiple pre-processor inclusion through use of
*               the  CPU module present pre-processor macro definition.
*           (2) RT-Thread支持32位、64位CPU,uCOS-III支持8位、16位、32位但不支持64位CPU,因此当且仅当CPU为
*               32位时,才能将原有项目从uCOS-III兼容到RT-Thread上.
*               当然,如果想在64位CPU上使用uCOS-III接口编程也是可以的(因为底层CPU调用的是RTT的接口)
*********************************************************************************************************
*/

#ifndef  CPU_MODULE_PRESENT                                     /* See Note #1.                                         */
#define  CPU_MODULE_PRESENT

#ifdef __cplusplus
extern  "C" {
#endif

#include <rtdef.h>
#include "cpu_def.h"
#include <rtconfig.h>

/*
*********************************************************************************************************
*                                    CONFIGURE STANDARD DATA TYPES
*
* Note(s) : (1) Configure standard data types according to CPU-/compiler-specifications.
*
*           (2) (a) (1) 'CPU_FNCT_VOID' data type defined to replace the commonly-used function pointer
*                       data type of a pointer to a function which returns void & has no arguments.
*
*                   (2) Example function pointer usage :
*
*                           CPU_FNCT_VOID  FnctName;
*
*                           FnctName();
*
*********************************************************************************************************
*/

typedef            void        CPU_VOID;
typedef            char        CPU_CHAR;                        /*  8-bit character                                     */
typedef       rt_bool_t        CPU_BOOLEAN;                     /*  8-bit boolean or logical                            */
typedef      rt_uint8_t        CPU_INT08U;                      /*  8-bit unsigned integer                              */
typedef       rt_int8_t        CPU_INT08S;                      /*  8-bit   signed integer                              */
typedef     rt_uint16_t        CPU_INT16U;                      /* 16-bit unsigned integer                              */
typedef      rt_int16_t        CPU_INT16S;                      /* 16-bit   signed integer                              */
typedef     rt_uint32_t        CPU_INT32U;                      /* 32-bit unsigned integer                              */
typedef      rt_int32_t        CPU_INT32S;                      /* 32-bit   signed integer                              */
typedef     rt_uint64_t        CPU_INT64U;                      /* 64-bit unsigned integer                              */
typedef      rt_int64_t        CPU_INT64S;                      /* 64-bit   signed integer                              */

typedef            float       CPU_FP32;                        /* 32-bit floating point                                */
typedef            double      CPU_FP64;                        /* 64-bit floating point                                */


typedef  volatile  CPU_INT08U  CPU_REG08;                       /*  8-bit register                                      */
typedef  volatile  CPU_INT16U  CPU_REG16;                       /* 16-bit register                                      */
typedef  volatile  CPU_INT32U  CPU_REG32;                       /* 32-bit register                                      */
typedef  volatile  CPU_INT64U  CPU_REG64;                       /* 64-bit register                                      */


/*
*********************************************************************************************************
*                                       CPU WORD CONFIGURATION
*
* Note(s) : (1) Configure CPU_CFG_ADDR_SIZE, CPU_CFG_DATA_SIZE, & CPU_CFG_DATA_SIZE_MAX with CPU's &/or
*               compiler's word sizes :
*
*                   CPU_WORD_SIZE_08             8-bit word size
*                   CPU_WORD_SIZE_16            16-bit word size
*                   CPU_WORD_SIZE_32            32-bit word size
*                   CPU_WORD_SIZE_64            64-bit word size
*
*           (2) Configure CPU_CFG_ENDIAN_TYPE with CPU's data-word-memory order :
*
*               (a) CPU_ENDIAN_TYPE_BIG         Big-   endian word order (CPU words' most  significant
*                                                                         octet @ lowest memory address)
*               (b) CPU_ENDIAN_TYPE_LITTLE      Little-endian word order (CPU words' least significant
*                                                                         octet @ lowest memory address)
*********************************************************************************************************
*/

                                                                /* Define  CPU         word sizes (see Note #1) :       */
#define  CPU_CFG_ADDR_SIZE              CPU_WORD_SIZE_32        /* Defines CPU address word size  (in octets).          */
#define  CPU_CFG_DATA_SIZE              CPU_WORD_SIZE_32        /* Defines CPU data    word size  (in octets).          */
#define  CPU_CFG_DATA_SIZE_MAX          CPU_WORD_SIZE_64        /* Defines CPU maximum word size  (in octets).          */

#define  CPU_CFG_ENDIAN_TYPE            CPU_ENDIAN_TYPE_LITTLE  /* Defines CPU data    word-memory order (see Note #2). */


/*
*********************************************************************************************************
*                                 CONFIGURE CPU ADDRESS & DATA TYPES
*********************************************************************************************************
*/

typedef  rt_ubase_t  CPU_ADDR;                                  /* CPU address type based on address bus size.          */
typedef  rt_ubase_t  CPU_DATA;                                  /* CPU data    type based on data    bus size.          */

typedef  CPU_DATA    CPU_ALIGN;                                 /* Defines CPU data-word-alignment size.                */
typedef  CPU_ADDR    CPU_SIZE_T;                                /* Defines CPU standard 'size_t'   size.                */

/*
*********************************************************************************************************
*                                       CPU STACK CONFIGURATION
*
* Note(s) : (1) Configure CPU_CFG_STK_GROWTH in 'cpu.h' with CPU's stack growth order :
*
*               (a) CPU_STK_GROWTH_LO_TO_HI     CPU stack pointer increments to the next higher  stack
*                                                   memory address after data is pushed onto the stack
*               (b) CPU_STK_GROWTH_HI_TO_LO     CPU stack pointer decrements to the next lower   stack
*                                                   memory address after data is pushed onto the stack
*
*           (2) Configure CPU_CFG_STK_ALIGN_BYTES with the highest minimum alignement required for
*               cpu stacks.
*
*               (a) ARM Procedure Calls Standard requires an 8 bytes stack alignment.
*********************************************************************************************************
*/

#if defined(ARCH_CPU_STACK_GROWS_UPWARD)                        /* Defines CPU stack growth order (see Note #1).        */
#define  CPU_CFG_STK_GROWTH     CPU_STK_GROWTH_LO_TO_HI
#else
#define  CPU_CFG_STK_GROWTH     CPU_STK_GROWTH_HI_TO_LO
#endif

typedef  rt_ubase_t             CPU_STK;                        /* Defines CPU stack word size (in octets).             */
typedef  CPU_ADDR               CPU_STK_SIZE;                   /* Defines CPU stack      size (in number of CPU_STKs). */

#define  CPU_CFG_STK_ALIGN_BYTES  (8u)                          /* Defines CPU stack alignment in bytes. (see Note #2). */

/*
*********************************************************************************************************
*                                   CRITICAL SECTION CONFIGURATION
*
* Note(s) : (1) (a) To save/restore interrupt status, a local variable 'cpu_sr' of type 'CPU_SR' MAY need
*                   to be declared (e.g. if 'CPU_CRITICAL_METHOD_STATUS_LOCAL' method is configured).
*
*                   (1) 'cpu_sr' local variable SHOULD be declared via the CPU_SR_ALLOC() macro which, if
*                        used, MUST be declared following ALL other local variables.
*
*                        Example :
*
*                           void  Fnct (void)
*                           {
*                               CPU_INT08U  val_08;
*                               CPU_INT16U  val_16;
*                               CPU_INT32U  val_32;
*                               CPU_SR_ALLOC();         MUST be declared after ALL other local variables
*                                   :
*                                   :
*                           }
*
*               (b) Configure 'CPU_SR' data type with the appropriate-sized CPU data type large enough to
*                   completely store the CPU's/compiler's status word.
*********************************************************************************************************
*/

typedef  rt_ubase_t                 CPU_SR;                     /* Defines   CPU status register size (see Note #3b).   */

#define  CPU_SR_ALLOC()             CPU_SR  cpu_sr = (CPU_SR)0
                                                                        /* Save    CPU status word & disable interrupts.*/
#define  CPU_INT_DIS()         do { cpu_sr = CPU_SR_Save();} while (0)
#define  CPU_INT_EN()          do { CPU_SR_Restore(cpu_sr); } while (0) /* Restore CPU status word.                     */


#define  CPU_CRITICAL_ENTER()  do { CPU_INT_DIS(); } while (0)          /* Disable   interrupts.                        */
#define  CPU_CRITICAL_EXIT()   do { CPU_INT_EN();  } while (0)          /* Re-enable interrupts.                        */


/*
*********************************************************************************************************
*                                         FUNCTION PROTOTYPES
*********************************************************************************************************
*/

CPU_SR      CPU_SR_Save      (void);
void        CPU_SR_Restore   (CPU_SR      cpu_sr);

#ifdef __cplusplus
}
#endif

#endif                                                          /* End of CPU module include.                           */
